dataverse/webapi/C#-NETx/BulkOperations/UpsertMultiple/Program.cs

```diff
@@ -0,0 +1,349 @@
+                            { "@odata.id", $"{tableLogicalName}s({tableAlternateKey}='sample record upsert record key {i+1:0000000}')" },
+                            // Each record MUST have the @odata.type specified.
+                            {"@odata.type",$"Microsoft.Dynamics.CRM.{tableLogicalName}" }
+                        });
+                            }
+                        }
+                    }
+
+                    // Use UpsertMultipleRequest
+                    UpsertMultipleRequest upsertMultipleRequest2 = new(
+                        entitySetName: tableSetName,
+                        targets: entityList);
+
+                    // Add a tag optional parameter to set a shared variable to be available to a plug-in.
+                    // See https://learn.microsoft.com/power-apps/developer/data-platform/optional-parameters?tabs=webapi#add-a-shared-variable-to-the-plugin-execution-context
+
+                    // TODO?: Add a 'tag' parameter to the UpsertMultipleRequest constructor provide the option to set this
+                    // rather than updating the existing URL?
+                    // CreateMultiple and UpdateMultiple don't have this though...
+                    upsertMultipleRequest2.RequestUri = new Uri(
+                        uriString: upsertMultipleRequest2.RequestUri.ToString() + "?tag=UpsertMultiple", 
+                        uriKind: UriKind.Relative);
+
+                    if (Settings.BypassCustomPluginExecution)
+                    {
+#pragma warning disable CS0162 // Unreachable code detected: Configurable by setting
+                        upsertMultipleRequest2.Headers.Add("MSCRM.BypassCustomPluginExecution", "true");
+#pragma warning restore CS0162 // Unreachable code detected: Configurable by setting
+                    }
+
+                    Console.WriteLine($"Sending POST request to /{tableSetName}/Microsoft.Dynamics.CRM.UpsertMultiple");
+
+                    Stopwatch upsertStopwatch = Stopwatch.StartNew();
+                    // Send the request
+                    await service.SendAsync(upsertMultipleRequest2);
+                    upsertStopwatch.Stop();
+                    Console.WriteLine($"\tUpserted {entityList.Count} records " +
+                        $"in {Math.Round(upsertStopwatch.Elapsed.TotalSeconds)} seconds.");
+
+                    // Delete created rows asynchronously
+                    // When testing plug-ins, set break point here to observe data
+                    Console.WriteLine($"\nStarting asynchronous bulk delete " +
+                        $"of {entityList.Count} upserted records...");
+
+                    var retrieveRequest = new RetrieveMultipleRequest(queryUri: retrieveMultipleQuery);
+
+                    var retrieveMultipleResponse2 = await service.SendAsync<RetrieveMultipleResponse>(retrieveRequest);
+                    // TODO: RetrieveMultipleResponse.Records contains a JArray of JObjects
```

Comment:
As mentioned above, You should be able to use the RetrieveMultipleResponse.Records property